Frontend production update

linearcombination opened the pull request:
This is a big update which incorporates everyone's commits and accomplishes many desired goals of devops and development. Please read commit logs for details as they have been carefully written.

linearcombination committed:
Initial commit of frontend

linearcombination committed:
lang0, lang1, lang2 languages working end to end.

Each including resource types and resource codes and handling backend
error reporting.

linearcombination committed:
Remove unneeded assets and there references

linearcombination committed:
Add needed binaries for node to install packages

Some packages require the presence of Python and g++

linearcombination committed:
Working Docker build with nginx config

We can now build the docker container successfully

linearcombination committed:
Handle success and failure properly.

linearcombination committed:
Better log messages

linearcombination committed:
Add CMD stanza to start frontend

linearcombination committed:
Reorder a few commands to max out build caching

Change order of nginx related commands to maximize docker build caching

linearcombination committed:
Remove now irrelevant source comments

danparisd committed:
Create docker.yml

linearcombination committed:
Remove git submodule for frontend

...going to monorepo to more easily comport with production build
system.

linearcombination committed:
Serve built assets with nginx (#2)

* Enable two stage Docker build

Stage 0 builds the project assets, stage 1 copies these assets into an
nginx environment where it will be served.

* Remove obsolete source comment

danparisd committed:
add dev backend CORS

danparisd committed:
80 for nginx

danparisd committed:
remove events

danparisd committed:
fix server

linearcombination committed:
Make backend API URL configurable via .env file

The value in .env for SNOWPACK_PUBLIC_BACKEND_PROD_API_URL will be
used when the snowpack deployment mode is 'production' which it will
be when 'snowpack build' is invoked via 'npm run build' (due to config
in package.json). Otherwise a value for API URL is used which is
compatible with local development, i.e., when snowpack deployment mode
is 'development' which it is when 'snowpack dev' is invoked via 'npm
start' (due to config in package.json).

linearcombination committed:
Update README

More detail

linearcombination committed:
Minor improvement to source comment

linearcombination committed:
Add back in CMD to launch snowpack in development mode as option

See source code comment for details on how this is used.

danparisd committed:
set location

danparisd committed:
better nginx conf

danparisd committed:
fix missing }

Commit:
fix static file location

danparisd committed:
Update .env

linearcombination committed:
Move to monorepo with frontend changes too

* Change to monorepo

We are no longer using a separate repo for the frontend and
associated submodule. All commits have been replayed onto the monorepo
to preserve everyone's work and additional changes have been made
related to upgrades.

* Change from snowpack to vite

The svelte community has abandoned snowpack in favor of the vite build
and runtime system. This change improved the system both in terms of
build speed and also solved a problem wherein use of environment
variables in snowpack was broken.

* Frontend now uses a string table in the form of .env and .env.production

Vite provides the ability to inject environment variables defined in
.env and .env.production (See
https://vitejs.dev/guide/env-and-mode.html#env-variables).
./frontend/src/App.svelte has been updated to use this feature. Now
all text that is displayed in the user interface can be changed by
programmers or non-programmers by changing the values in .env and
.env.production and then doing a 'docker-compose build' or
equivalently a 'make build'.

linearcombination committed:
Better organize dependencies in Makefile

Keep things DRY by better utilizing Makefile target dependencies.

linearcombination committed:
Minor source comment formatting

linearcombination committed:
Remove unused code and comments

Remove old legacy cruft of code that was maybe going to be used, but
never was.

linearcombination committed:
Update BACKEND_CORS_ORIGINS

...for change to vite build and runtime system. Remove any superfluous
entries in array and add any that are necessary.

linearcombination committed:
Remove unnecessary quotes around port mappings

...quotes are not needed.

linearcombination committed:
One docker-compose.yml for development and production

Fulfill principle X of 12factorapp: https://12factor.net/dev-prod-parity

linearcombination committed:
Update Makefile to work with docker-compose.yml changes

backend renamed to api

linearcombination committed:
Remove unnecessary --force-recreate switch

...after more experience with Docker I've found I don't need this with
the new workflow I have adopted.

linearcombination committed:
Improve source comment

danparisd commented on .env:
```diff
@@ -60,8 +53,17 @@ ENABLE_ASSET_CACHING=true
+#  * https://irguidev.walink.org covers requests originating from the
+#  development/staging server for the (G)UI at WA.
+#  (Array must be JSON formatted)
+BACKEND_CORS_ORIGINS='["http://localhost:3000", "http://localhost:8001", "https://irguidev.walink.org"]'
```
```suggestion
BACKEND_CORS_ORIGINS='["http://localhost:3000", "http://localhost:8001"]'
```
we can leave it like this and set the dev/prod CORS origins via environment variable. I just did this as a hack.

@linearcombination can you make this change so I can merge the PR?

linearcombination replied:
This is complete. Thank you.

linearcombination committed:
Prefer local reasoning over global reasoning

Make functions less reliant on global or module level constants. Instead
pass them in as defaulted arguments. We'd use functools.partial instead
of default arguments, but the partial decorator interferes with the
ability to compile the code using mypyc, otherwise we would use partial
as that would be better FP style.

In more formal terms: further increase referential transparency [1] via
use of function parameter scope dependency injection. [2]

[1] https://en.wikipedia.org/wiki/Referential_transparency
[2] https://www.toptal.com/python/python-parameterized-design-patterns

linearcombination committed:
Minor source comment updates

linearcombination committed:
Simplify type

linearcombination committed:
Remove now obsolete file

..this file references a package that was used in the paste to do
wholesale type hint upgrade across multiple files. No longer used.

linearcombination committed:
Strong (strict) typing in Typescript

...strong typing in Typescript, like same in Python via mypy, is a
great way to ensure higher code quality and better system
documentation. Adding a type check as part of the frontend Docker
build ensures that future changes will at least stay strongly typed.

Add loading indicator child component. Note that we use a css class to
hide the first language list while it is loading rather than using a
svelte #if statement so that we can start loading the first language
list in the background while the user is interacting with the assembly
strategies drop down. This shortens the perceived delay for the first
language list to load.

Move assembly strategy into child component.

Configure javascript and svelte linting. Resulting code formatting.

Better organization of javascript code.

Abandon attempt at LanguageComponent for now as it had a weird bug and
actually makes the code less understandable.

linearcombination committed:
Fix bug in helps column rendering

Logic was simplified which eliminated the bug. Added test case that
triggered the bug, though the test itself did not detect the issue as
written, but did generate the PDF that had the column rendering issue
which had to be viewed manually.

linearcombination committed:
Update BACKEND_CORS_ORIGINS for production

One of the origins is not needed in .env because it will be set via
env var in production.

linearcombination committed:
Mkinor cleanup of unnecessary variables

linearcombination committed:
Upgrade node docker image

linearcombination committed:
Upgrade nginx docker image

linearcombination committed:
Fix comment
